### PR TITLE
[#30] [API] As a user, I can upload a CSV file of keyword

### LINKED ISSUE #30: [API] As a user, I can upload a CSV file of keyword
## Why

Create an API that a user can use upload CSV file.

## Acceptance Criteria

The API must meet the following conditions:
- JSON API standard must be followed
- Use the same code for validation and process CSV field that we used on the backend: https://github.com/Nihisil/nimble-certification-elixir/issues/10

### PR DESCRIPTION
**Pull request by Nihisil**
Resolves #30

## What happened 👀

New API endpoint to upload keyword file was added.

## Proof Of Work 📹

image

### CODE REVIEW

File: lib/google_scraping_web/controllers/api/v1/upload_keyword_controller.ex [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+  def create(conn, _params) do
+    ErrorHandler.render_error_json(
+      conn,
+      :unprocessable_entity,
+      "Invalid input attributes. Add `keyword_csv_file` to request body"
+    )
+  end
+
+  defp process_validation_error(conn, reason) do
```
Comment (junan) [inline lib/google_scraping_web/controllers/api/v1/upload_keyword_controller.ex:47]: It it possible to move this function in a separate module? as this one is quite big and makes the controller fat. may be in `ErrorHandler` or other separate module?

File: lib/google_scraping_web/controllers/error_handler.ex [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+  def build_changeset_error_message(changeset) do
+    errors =
+      Changeset.traverse_errors(changeset, fn {msg, opts} ->
+        Enum.reduce(opts, msg, fn {key, value}, acc ->
+          String.replace(acc, "%{#{key}}", to_string(value))
+        end)
+      end)
+
+    error_message =
```
Comment (junan) [inline lib/google_scraping_web/controllers/error_handler.ex:34]: As we are not using this variable(`error_message `) anywhere, can we just return `Enum.map_join(errors, "\n", fn {key, errors} -> "#{key} #{Enum.join(errors, ", ")}" end)` directly

File: test/google_scraping_web/api/v1/upload_keyword_controller_test.exs [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+    assert %{
+             "errors" => [
+               %{
+                 "detail" => "The file is too big, allowed size is up to 1000 keywords",
+                 "status" => "bad_request"
+               }
+             ]
+           } ==
+             json_response(conn, 400)
```
Comment (junan) [inline test/google_scraping_web/api/v1/upload_keyword_controller_test.exs:66]: I think it should return `422` as this this validation error(submitted params value is not valid)

File: test/google_scraping_web/api/v1/upload_keyword_controller_test.exs [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+  end
+
+  test "given a non valid file, returns 400 status with error details", %{conn: conn} do
```
Comment (junan) [inline test/google_scraping_web/api/v1/upload_keyword_controller_test.exs:69]:
```suggestion
  test "given a NON valid file, returns 400 status with error details", %{conn: conn} do
```

File: test/google_scraping_web/api/v1/upload_keyword_controller_test.exs [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+    assert %{
+             "errors" => [
+               %{
+                 "detail" => "One or more keywords are invalid! Allowed keyword length is 1-100",
+                 "status" => "bad_request"
+               }
+             ]
+           } ==
+             json_response(conn, 400)
```
Comment (junan) [inline test/google_scraping_web/api/v1/upload_keyword_controller_test.exs:88]: Same as this comment https://github.com/Nihisil/nimble-certification-elixir/pull/55/files#r971483851

File: lib/google_scraping_web/controllers/api/v1/upload_keyword_controller.ex (hunk 1 above)
Comment (Nihisil) [reply lib/google_scraping_web/controllers/api/v1/upload_keyword_controller.ex:47]: I'm not sure that it is worth moving full error handling to a separate module. It is directly related to the logic of the controller - validation of user input.

I refactored this method a bit. Now it looks better: https://github.com/Nihisil/nimble-certification-elixir/pull/55/commits/0ec08506524cbb09538bb112231e1319fffb292f

File: lib/google_scraping_web/controllers/error_handler.ex (hunk 2 above)
Comment (Nihisil) [reply lib/google_scraping_web/controllers/error_handler.ex]: Was fixed here: https://github.com/Nihisil/nimble-certification-elixir/pull/55/commits/ff6f9aeb36ecfcdb0b90e675fa6768de6f08a432

File: test/google_scraping_web/api/v1/upload_keyword_controller_test.exs (hunk 5 above)
Comment (Nihisil) [reply test/google_scraping_web/api/v1/upload_keyword_controller_test.exs]: Yep, I agree. It was fixed here: https://github.com/Nihisil/nimble-certification-elixir/pull/55/commits/a6bc2fd9261ed86349ce3e2862a91b7f6d43bfde

File: test/google_scraping_web/api/v1/upload_keyword_controller_test.exs [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+    upload_file = keyword_file_fixture("empty.csv")
+
+    conn =
+      conn
+      |> token_auth_user(user)
+      |> post(Routes.api_upload_keyword_path(conn, :create), %{
+        keyword_csv_file: upload_file
+      })
+
+    assert %{"errors" => [%{"detail" => "The file is empty", "status" => "unprocessable_entity"}]} ==
```
Comment (junan) [inline test/google_scraping_web/api/v1/upload_keyword_controller_test.exs:43]: We keep the matching value at the right side when we assert with `==` to make it more redable

so it should be `assert json_response(conn, 422) ==  %{"errors" ...`

File: test/google_scraping_web/api/v1/upload_keyword_controller_test.exs [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+    upload_file = keyword_file_fixture("big.csv")
+
+    conn =
+      conn
+      |> token_auth_user(user)
+      |> post(Routes.api_upload_keyword_path(conn, :create), %{
+        keyword_csv_file: upload_file
+      })
+
+    assert %{
```
Comment (junan) [inline test/google_scraping_web/api/v1/upload_keyword_controller_test.exs:58]: Same as above and others

**Review by junan — APPROVED** [review]

File: test/google_scraping_web/api/v1/upload_keyword_controller_test.exs (hunk 6 above)
Comment (Nihisil) [reply test/google_scraping_web/api/v1/upload_keyword_controller_test.exs]: Yes, thanks. I always forgot about this one.

**Review by bterone — APPROVED** [review]

**Pull request merged**